### ISSUE TITLE
Update CBBR data

### ISSUE THREAD

**Issue by AmandaDoyle** [NYCPlanning/db-cbbr#27]
Goal: Generate new CBBR data for Planning Support.  These data are used to inform capital investments across the city.

**Process for updating CBBR data**
- Create workflows in data library to archive source data in data library.  The last time CBBR was built was 2020; therefore, it's processes are out out date.
- Update other core infrastructure as needed for build.  Overhaul / full update of repo should be saved for a dedicated maintenance sprint.
- Update CBBR build workflow to load source data from data library
- Update build workflow to incorporate and output agency responses.  This is the first time we've received agency responses.
- Report any missing source data fields and/or input tables to Planning Support and AD
- For fields where the values in the fields changed (i.e. `BX 01` vs `201`) determine if it's necessary to change the values to match previous versions of CBBR, and if so determine if it's best to make these changes to the source data or in the code.
- Build CBBR
- Review outputs
- Delegate the manual mapping of projects across EDM (dependent on volume)
- Incorporate manually mapped data back into the CBBR build process
- Rerun CBBR with additional geospatial data (manually mapped projects)
- Review outputs
- Deliver to Planning Support

**Inputs**
See [here](https://github.com/NYCPlanning/db-cbbr/blob/master/cbbr_build/01_dataloading.sh), [here](https://github.com/NYCPlanning/db-cbbr/blob/master/cbbr_build/python/dataloading.py)
- cbbr_submissions --> CBBR_Registar_Extract tab in CBBR_OMBextractFY24 (Excel file from NS)
- cbbr_agency_updates - I think this will be deprecated since agency responses are in main table.
- dpr_parksproperties - in data library
- doitt_buildingfootprints - in data library
- facilities - this is dcp_facilities in data library aka FacDB.  Should we wait to build CBBR until after FacDB?
- old_cbbr_submissions - I think this would be the previous version of CBBR, which should be in Carto, but we'd need to track down if we determine we need it
- Manually mapped geometries, which are stored in various formats [here](https://github.com/NYCPlanning/db-cbbr/tree/master/cbbr_build/geometries).

**Outputs**
See [here](https://github.com/NYCPlanning/db-cbbr#available-files)

- cbbr_submissions.csv
- cbbr_submissions_needgeoms.csv
- cbbr_submissions_poly.csv
- cbbr_submissions_pts.csv
- cbbr_submissions_poly.zip
- cbbr_submissions_pts.zip

**Comment by damonmcc** [comment]
Max is starting on the input side (e.g. adding data library templates)

I'm starting on the output side (e.g. ensuring build process aligns with expectations and other data products)

**Comment by mbh329** [comment]
Adding data library templates issue [here](https://github.com/NYCPlanning/db-data-library/issues/374)

**Comment by damonmcc** [comment]
noted a weird thing with `doitt_buildingfootprint` data in [this PR comment](https://github.com/NYCPlanning/db-cbbr/pull/28#issuecomment-1432306923). temporary fix is to pin the version to the last one with a "valid" sql file in the edm-recipes S3 bucket

**Comment by damonmcc** [comment]
status update
- the build is generally working
- we have new source data (emailed) which has the expected columns `type` and `need`. `type` is especially important for selecting records we actually want to manually map after the first build
- as significant number of records fail geocoding because the address data (if there is any) is in the `location` column, which was not the case last time

**Comment by mbh329** [comment]
@damonmcc data is updated in edm-recipes with the additional field names `type` and `need`

**Comment by damonmcc** [comment]
status update
* latest data with new columns
* geocoding now parses `location` column but most records don't geocode and aren't sites (screenshots below of queries on `_cbbr_submissions` table aka `cbbr_submissions_build.csv`)
<img width="251" alt="image" src="https://user-images.githubusercontent.com/7444289/220475915-fa1b30f8-7665-4f48-a364-e15d7873b4c7.png">
<img width="291" alt="image" src="https://user-images.githubusercontent.com/7444289/220476169-8846667f-da91-4faf-a1e0-9d37bac4864c.png">
<img width="492" alt="image" src="https://user-images.githubusercontent.com/7444289/220476300-253c6cfc-cae2-45ca-9274-7ad79932f600.png">

**Comment by damonmcc** [comment]
the borough codes in the input data are NOT the same as Geosupport 🙃

working on converting the borough names to the correct codes

cc @mbh329

**Comment by damonmcc** [comment]
current status of outputs:
- 3728 records in total
- 1290 records with a valid location (not null, not district wide)
- 720 records with spatial data (either from geocoding or matching to parks/facilities data)
- 706 records with a geometry
- **55% geometry hit rate (records with a geometry / records with a valid location)**

the primary dev branch is in review and will be merged soon. I'll work to implement geosupport functions 2 and 3 to increase geometries (ideally in a new PR)

**Comment by AmandaDoyle** [comment]
Notes from AD review 02/23 at noon.  No concerns or issues, just thoughts on improvement.
- Currently, polygons are only coming from the parks properties dataset.  Since now we're passing common place names through GeoSupport there are fewer records than before getting spatial data from parks properties.  Should CBBR be a points based database?  We can get the centroid of the park (but need to make sure that the point falls within the boundaries of polygon).  More projects be polygons once we implement function 3, but TBD how many and those could be just the point on the center of the street segment.  This is up for discussion within DE and should be discussed with EDM and PS.
- We may want to consider limiting `cbbr_submissions_needgeoms` to records `WHERE type = 'site' and type_br = 'C' or at least prioritize records where type_br = 'C'.  This gets it down to ~250ish records.  I do not think we can limit research to records where the `location` field is not null because Community Boards may not use fields the same.  For example "NYPL should winterize the wonderful outdoor space at Baychester Branch." is the explanation for a record with no location information.  
- As an enhancement should we try to categorize project like we do in CPDB (see [here](https://github.com/NYCPlanning/db-cpdb/blob/main/sql/projectscategorization.sql)).  This will help us focus on manually mapping projects that are not vehicles and IT equipment.
- Shapefiles cut off titles to 10 characters.  Should we modify the field names of the shapefiles to accommodate this?

**Comment by damonmcc** [comment]
current status of [outputs on latest dev branch](https://github.com/NYCPlanning/db-cbbr/tree/dm/improve-geometry-rate/cbbr_build/output):
- 3728 records in total
- 1307 records with a specified location (not null, not district wide)
- 939 records with a geometry
- 72% geometry hit rate (records with a geometry / records with a specified location)

records in need of geometries by priority level
- Priority A: 236
- Priority B: 962
- Priority C: 1592
- Total: 2790

**Comment by damonmcc** [comment]
@AmandaDoyle 
re: Nick Stampar email on 2/27:

> I'm a bit hesitant to exclude the expense requests outright. Do we know how many records would need reviewing if we included both capital and expense requests? 

2790 records


> On the polygon/point question, I do think it's reasonable to have them represented as points, and I'm sure this will satisfactorily represent the majority of CBBR locations, however the one exception we're concerned about is with any requests that reference a road segment, where the road may be quite long, potentially exiting across more than one borough, in which case the midpoint for the road may not actually be located near the segment of the request. Do we know how we would handle point-locating for road segments like this? 

A bit of clarification on Geosupport terms and what we were/are doing in geocoding:
- Street Segment 
  - A Street Segment is a single block length between two consective intersections. [Geosupport Function 3](https://a030-goat.nyc.gov/goat/Function3?borough1=3&borough2=3&borough3=3&streetName=Myrtle%20Ave&streetName2=Marcus%20Garvey%20BLVD&streetName3=Throop%20ave&compass=) handles these and provides the locations of the two ends of the segment. We create a straight line between the two intersections as a MultiPoint geometry.
  - We were using Function 3 before and are using it now.
- Street Stretch
  - A Street Stretch is a series of street segments. [Geosupport Function 3S](https://a030-goat.nyc.gov/goat/Function3S?borough=3&street1=Myrtle%20Ave&street2=Marcus%20Garvey%20BLV&street3=Tompkins%20Ave&compass1=&compass2=) handles these and provides the locations of the intersections and segments.
  - We weren't using Function 3S and aren't using it now (yet?).

**Comment by damonmcc** [comment]
status: manual mapping by EDM has begun

**Comment by mbh329** [comment]
Manually mapped Priority A projects have been incorporated into CBBR. We still need to update the geometries for Priority B projects which have not been scheduled. These will have to be done before we can publish a version of CBBR for Strategic Planning for FY2024. 

Publishing action works and a PR is [open](https://github.com/NYCPlanning/db-cbbr/pull/34)

Joining manually mapped projects (by `boro_and_board`, `request` and `location`) year over year by joining latest version of `cbbr_submissions` data with the previous years data has been scoped out in an [issue](https://github.com/NYCPlanning/db-cbbr/issues/36). I don't know how much sense it makes to incorporate any of this as the last version of CBBR data that we created was from FY2021 but the issue was fully scoped out and is viable based on the information we got from NS in Strategic Planning. We also don't know if the input data will change and, if so, how. I know there is talk scheduled to improve programatic geocoding on April 4 which will drastically reduce the number of records we will have to end up mapping (hopefully). Let me know if this makes sense. 

@AmandaDoyle @damonmcc @fvankrieken
